add decimal validator

davidhewitt opened the pull request:
## Change Summary

Adds a Rust implementation of decimal validation and serialization. Needs more tests and benchmarks to confirm it's worth going for this instead of https://github.com/pydantic/pydantic/pull/6327

## Related issue number

N/A

## Checklist

* [ ] Unit tests for the changes exist
* [ ] Documentation reflects the changes where applicable
* [ ] Pydantic tests pass with this `pydantic-core` (except for expected changes)
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**


Selected Reviewer: @dmontagu

samuelcolvin reviewed (dismissed):
looks like a good start.

The easiest way to check performance would be  to copy the validator constructed by pydantic and add it to benchmarks here, together with the function from pydantic. Then run benchmarks. You can probably remove that once we're sure it's faster.

samuelcolvin commented on src/input/input_python.rs:
```diff
@@ -347,6 +347,58 @@ impl<'a> Input<'a> for PyAny {
+
+        let str = if let Ok(str) = self.downcast::<PyString>() {
+            str
+        } else if let Ok(float) = self.downcast::<PyFloat>() {
+            float.str()?
+        } else if let Ok(int) = self.downcast::<PyInt>() {
+            int.str()?
```
given that we're just calling `.str()` and don't need the `PyFloat` / `PyInt`, it would probably be quicker to use `is_instance` - ideally we should use the `ObType` logic here so we only need to lookup the type pointer once.

samuelcolvin commented on src/input/input_python.rs:
```diff
@@ -347,6 +347,58 @@ impl<'a> Input<'a> for PyAny {
+        } else if let Ok(int) = self.downcast::<PyInt>() {
+            int.str()?
+        } else {
+            self.str().map_or(self, |s| s.as_ref())
```
does this logic really match pydantic? If so, we can remove the isinstance checks and just go straight to `.str()`, but I think we're stricter than this in pydantic.

davidhewitt replied:
I've reworked in a way which should match Pydantic.

samuelcolvin reviewed (commented):
couple more points.

davidhewitt commented on src/input/input_json.rs:
```diff
@@ -171,6 +172,27 @@ impl<'a> Input<'a> for JsonInput {
 
+    fn strict_decimal(&'a self, decimal_type: &Py<PyType>) -> ValResult<&'a PyAny> {
+        // FIXME pure evil!
+        let py = unsafe { Python::assume_gil_acquired() };
```
Yes that's exactly the solution, I was cutting corners to get a first implementation running ;)

samuelcolvin commented on src/validators/decimal.rs:
```diff
@@ -0,0 +1,254 @@
+    ) -> ValResult<'data, PyObject> {
+        // FIXME might need to check input type rather than blindly call, esp in strict mode
+        let decimal = input.validate_decimal(extra.strict.unwrap_or(self.strict), &self.decimal_type)?;
+        // FIXME use PyO3 raw ffi PyObject_CallOneArg ?
+        // let decimal = self.decimal_type.as_ref(py).call1((either_decimal,))?;
```
i guess this comment can be removed, if such a call is required, it should be within `validate_decimal` in the python implementation.

davidhewitt commented:
I think the implementation here for the validator is now complete and is as optimized as can be. Still to do:
- Serializer feedback
- More tests
- Create known error types

I will park this until we make a decision whether to move forwards with it. The alternative is https://github.com/pydantic/pydantic/pull/6781

JeanArhancet commented on src/validators/decimal.rs:
```diff
@@ -0,0 +1,255 @@
+    input: &'a impl Input<'a>,
+    decimal_type: &'a PyType,
+) -> ValResult<'a, &'a PyAny> {
+    decimal_type.call1((arg,)).map_err(|e| {
```
Concerning this part maybe we can do the calculation in rust and set the attributes of the python class `Decimal`: `__slots__ = ('_exp','_int','_sign', '_is_special')` WDYT @davidhewitt ?

davidhewitt replied:
In this PR I'm trying to keep the same logic as `pydantic` just move the implementation. I would definitely be interested in this as a future improvement. I believe @samuelcolvin had some previous thoughts on how the parsing might be done (by looking at the digits as a string), not sure if there was any implementation put somewhere?

davidhewitt commented:
please review

dmontagu commented on generate_self_schema.py:
```diff
@@ -23,6 +24,7 @@
     UnionType = Union[TypingUnionType, TypesUnionType]
 
 except ImportError:
+    TypesUnionType = None
```
Maybe better to do TypesUnionType = TypingUnionType, since there is an `or origin is TypesUnionType` below, where I don't think you want `origin is None` to be acceptable?

dmontagu replied:
```suggestion
    TypesUnionType = TypingUnionType
```

davidhewitt replied:
Thanks

dmontagu commented on generate_self_schema.py:
```diff
@@ -79,6 +81,7 @@ def get_schema(obj) -> core_schema.CoreSchema:
         # can't really use 'is-instance' since this is used for the class_ parameter of 'is-instance' validators
         return {'type': 'any'}
     else:
+        print(origin)
```
```suggestion
```

davidhewitt replied:
Thanks, yes a hangover from debugging.

dmontagu commented on python/pydantic_core/core_schema.py:
```diff
@@ -659,6 +660,72 @@ def float_schema(
+    le: Union[int, Decimal]
+    max_digits: int
+    decimal_places: int
+    multiple_of: Union[int, Decimal]
```
Should `float` be allowed here? I think we allow that with some tolerance elsewhere

davidhewitt replied:
Actually because `lax_decimal` will accept `int` and `float` there is no need for `Union` here, this should just be `Decimal`.

dmontagu commented on python/pydantic_core/core_schema.py:
```diff
@@ -659,6 +660,72 @@ def float_schema(
+    decimal_places: int
+    multiple_of: Union[int, Decimal]
+    allow_inf_nan: bool  # whether 'NaN', '+inf', '-inf' should be forbidden. default: False
+    check_digits: bool  # FIXME document. default: False
```
not sure what this is referring to, I guess I'll find out once it's documented

davidhewitt replied:
`check_digits` actually shouldn't be part of the schema, it's an internal state of the validator, will remove from here.

dmontagu commented on python/pydantic_core/core_schema.py:
```diff
@@ -659,6 +660,72 @@ def float_schema(
+    le: int | Decimal | None = None,
+    max_digits: int | None = None,
+    decimal_places: int | None = None,
+    multiple_of: int | Decimal | None = None,
```
looks like maybe this is missing `whole_digits`? I see it can be an error below but don't see a way to specify it

dmontagu replied:
After reading the validation code, it looks like whole_digits is basically `max_digits - decimal_places`. I didn't look closely but I wonder if it makes sense to allow these to be constrained separately? Not really a big deal I guess.

If this is essentially just copying the current python-side logic as faithfully as possible, I'm fine keeping it behaving as it currently does.

davidhewitt replied:
Yep it's a copy of the existing Python logic, I'd be open to considering a change in a future PR, let's keep as-is for now.

dmontagu commented on src/input/input_json.rs:
```diff
@@ -172,6 +173,18 @@ impl<'a> Input<'a> for JsonInput {
+    fn strict_decimal(&'a self, decimal_type: &'a PyType) -> ValResult<&'a PyAny> {
+        let py = decimal_type.py();
+        match self {
+            JsonInput::Float(f) => create_decimal(PyString::new(py, &f.to_string()), self, decimal_type),
```
Is there a way for us to make sure that when parsing numeric JSON into a decimal it is treated as a string rather than loaded into an f64 or whatever? I think @samuelcolvin said he did something like this in speedate

dmontagu replied:
(Maybe it's already doing that, not sure, but it's probably worth checking that we load many-digit JSON numbers into decimal properly)

dmontagu commented on src/input/input_python.rs:
```diff
@@ -352,6 +354,56 @@ impl<'a> Input<'a> for PyAny {
+            return Ok(self);
+        }
+
+        // Try subclasses of decimals, they will be upcast to Decimal
```
Any reason to not allow instances of subclasses of decimal through without an upcast?

davidhewitt replied:
In general this is what we've decided as the treatment of subclasses for V2 (e.g. int, float subclasses also upcast). In a future release maybe we will change this.

dmontagu commented on src/input/input_python.rs:
```diff
@@ -352,6 +354,56 @@ impl<'a> Input<'a> for PyAny {
+        ))
+    }
+
+    fn lax_decimal(&'a self, decimal_type: &'a PyType) -> ValResult<&'a PyAny> {
+        // Fast path for existing decimal objects
+        if self.is_exact_instance(decimal_type) {
+            return Ok(self);
+        }
+
+        if self.is_instance_of::<PyString>() || (self.is_instance_of::<PyInt>() && !self.is_instance_of::<PyBool>()) {
+            create_decimal(self, self, decimal_type)
+        } else if self.is_instance_of::<PyFloat>() {
+            create_decimal(self.str()?, self, decimal_type)
+        } else {
+            Err(ValError::new(ErrorType::DecimalType, self))
+        }
+    }
+
```
It looks to me like this will error for instances of subclasses of Decimal, not just upcast them.

But I don't see downside to just allowing proper subclass instances through without modification in strict _or_ lax mode. Maybe I'm missing something?

davidhewitt replied:
Good catch, thanks. I've now fixed the logic to allow subclasses here.

As per https://github.com/pydantic/pydantic-core/pull/763#discussion_r1285650677 I will upcast them for now.

dmontagu commented on tests/serializers/test_decimal.py:
Seems to me worth adding a test that `Decimal('123456789123456789123456789.123456789123456789123456789')` (or some other similarly high-precision decimal) gets serialized properly (at least to JSON).

dmontagu commented:
I made various comments; I think at least _some_ merit changes (e.g., removing the print line), and I do think there are some logic changes worth making (i.e., allowing proper subclasses of `Decimal` to pass validation unmodified), but happy to defer to your judgement.

For hooky: please update

davidhewitt commented on tests/validators/test_decimal.py:
I added a test to `test_decimal`, seems to work 👍

davidhewitt commented:
Thanks for the thorough review; that definitely caught a couple of bugs! 

I've rebased this PR and added a commit to make the desired review changes.

davidhewitt commented:
please review

davidhewitt commented:
Many thanks, I rebased and dropped the commit which pointed the integration tests at https://github.com/pydantic/pydantic/pull/6858 👍
